### DISCUSSION TITLE
崩坏：星穹铁道 抽卡记录导入导出格式

### DISCUSSION THREAD

**Discussion by LaoshuBaby** [UIGF-org/.github#7] [General]
目前收集到的项目有如下：

* [DancingSnow0517/StarRail-gacha](https://github.com/DancingSnow0517/StarRail-gacha/issues/4)
* [TremblingMoeNew/StarRailWarpObserve](https://github.com/TremblingMoeNew/StarRailWarpObserve/issues/1)
* [Scighost/StarRailTool](https://github.com/Scighost/StarRailTool/issues/4)
* [cntvc/star-rail-tools](https://github.com/cntvc/star-rail-tools/issues/4)
* [vikiboss/star-rail-gacha-export](https://github.com/vikiboss/star-rail-gacha-export/issues/1)
* [lgou2w/HoYo.Gacha](https://github.com/lgou2w/HoYo.Gacha/issues/6)
* [BoxCatTeam/SRCat](https://github.com/BoxCatTeam/SRCat/issues/5)
* [biuuu/star-rail-warp-export](https://github.com/biuuu/star-rail-warp-export/issues/16)

**Comment by TremblingMoeNew** [comment]
~~熟人？但怎么某游群里找不到你了~~
要搞统一格式的话，我的建议是直接放弃Excel导入，只用JSON。
Excel导出是用来给人看的，不是给程序看的

StaRWO（咦这么缩写挺顺口的，以后就叫它STARWO了.jpg）目前设计的导出格式（还没Release）大致基于UIGF.J V2.1，`info`里增加`game`（值仅为`Honkai_Star_Rail`）、`game_biz`（`hkrpg_cn`和`hkrpg_global`）以及`region_time_zone`，记录项那边就是敲掉`uid`和`lang`的原始数据。`uigf_gacha_type`的类似物还没加上，最终名字不确定嘛。

**Comment by DancingSnow0517** [comment]
uigf_gacha_type我觉得直接用原来的字符串数字？1，2，11，12？
以及我认为uid可以保留

**Comment by TremblingMoeNew** [comment]
> uigf_gacha_type我觉得直接用原来的字符串数字？1，2，11，12？ 

问题在“uigf”前缀不在值上(

毕竟只是给自己的软件整的导出格式，uigf毕竟目前只是个原神的祈愿记录导出格式标准

> 以及我认为uid可以保留

`info`里有`uid`和`lang`啊

**Comment by LaoshuBaby** [comment]
> 问题在“uigf”前缀不在值上(
> 
> 毕竟只是给自己的软件整的导出格式，uigf毕竟目前只是个原神的祈愿记录导出格式标准

刚刚在 https://github.com/lgou2w/HoYo.Gacha/issues/6 中有提到说，UIGF其实已经有标准组织了（因此我这个草台班子可以撤掉了233，可喜可贺），他们有在考虑支持星铁的抽卡记录了

Uniformed Interchangeable GachaLog Format standard (UIGF) 统一可交换成就标准 **和** Uniformed Interchangeable Achievement Format standard (UIAF) 统一可交换成就标准，**幸亏这里面的G不是Genshin**（原来UIAF也是他们搞出来的啊，之前有听说过但是不知道是哪一家在推动）

~~（不过我看那样UIGF组织也该改名了，毕竟还有UIAF）~~

**Comment by Lightczx** [comment]
我拉下组织，麻烦作者把issue 转移到UIGF吧
@LaoshuBaby

**Comment by LaoshuBaby** [comment]
> 我拉下组织，麻烦作者把issue 转移到UIGF吧
> @LaoshuBaby

感谢，正有此意（我直接整个仓库送过去好了XD）

**Comment by TremblingMoeNew** [comment]
> > 问题在“uigf”前缀不在值上(
> > 毕竟只是给自己的软件整的导出格式，uigf毕竟目前只是个原神的祈愿记录导出格式标准
> 
> 刚刚在 [lgou2w/HoYo.Gacha#6](https://github.com/lgou2w/HoYo.Gacha/issues/6) 中有提到说，UIGF其实已经有标准组织了（因此我这个草台班子可以撤掉了233，可喜可贺），他们有在考虑支持星铁的抽卡记录了
> 
> Uniformed Interchangeable GachaLog Format standard (UIGF) 统一可交换成就标准 **和** Uniformed Interchangeable Achievement Format standard (UIAF) 统一可交换成就标准，**幸亏这里面的G不是Genshin**（原来UIAF也是他们搞出来的啊，之前有听说过但是不知道是哪一家在推动）
> 
> ~（不过我看那样UIGF组织也该改名了，毕竟还有UIAF）~

UIGF格式的G是Gacha Log，但UIGF组织的G是Genshin……

**Comment by Lightczx** [comment]
@LaoshuBaby Check your email-box

**Comment by Lightczx** [comment]
折腾完了，成功变成讨论了

**Comment by TremblingMoeNew** [comment]
我的看法：

1. 仅保留JSON格式标准，放弃用Excel进行数据交换
2. `info`部分和UIGF.J v2.1一样放上`uid`、`lang`、`export_time`、`export_timestamp`、`export_app`、`export_app_version`，以及一个标识协议版本号的`uigf_version`。标记用户时区的`region_time_zone`有加入的必要。个人建议把`game_biz`也加进去。
4. 跃迁记录项就还是除`uid`和`lang`外的原始数据，`gacha_id`，`gacha_type`，`item_id`、`count`、`time`、`name`、`item_type`，`rank_type`，`id`，**这几项全部保留原始数据，不作任何改动**。和UIGF.J v2.1一样向记录项中注入`uigf_gacha_type`，以备双池子之需。如果要和UIGF.J v2.3一样考虑`item_type`的国际化问题，我认为应该注入新的`uigf_item_type`属性而不是修改原始数据，或者在UIGF翻译API里直接把item_id对应的item_type发下去不就完了
5. 有在`info`部分加入游戏标识以及协议标识的必要（其中`uigf_version`已经在一定程度上起到了后者的作用）。万一有人拿原的UIGF甚至UIAF往星铁的工具里导入呢2333333，而且也需要考虑同时兼容两个游戏的工具的识别问题。加个`game`还是很有必要的。

@Lightczx  @DancingSnow0517 @biuuu @Scighost @cntvc @vikiboss @lgou2w @Masterain98

**Reply by Masterain98** [reply]
直接按照 UIGF2.3 来改，否则后面多语言还是一团糟

**Reply by TremblingMoeNew** [reply]
> 直接按照 UIGF2.3 来改，否则后面多语言还是一团糟

2.3改了`item_id`和`item_type`的原始数据，我认为没有做到最大化保留原始数据以及最大化兼容。

星铁本身就提供了`item_id`。`item_type`的问题靠翻译API或者`uigf_item_type`都是可以的啊。

**Reply by Lightczx** [reply]
原始数据是可以保留的，UIGFv2.3只是要求设置item_id的值

**Reply by Lightczx** [reply]
我更加倾向于新开一个spec而不是升级UIGF

**Reply by TremblingMoeNew** [reply]
> 我更加倾向于新开一个spec而不是升级UIGF

我也支持为星铁重新写一版，不管这一版的名字到底叫啥。但原神UIGF和这个在导入过程中的区分标识我觉得还是得有的。

换个名字的话，原神的UIGF已经把名字给占了，叫啥是个问题（）

**Reply by Lightczx** [reply]
> 1. 仅保留JSON格式标准，放弃用Excel进行数据交换

反正我已经放弃了



> 3\. 和UIGF.J v2.1一样向记录项中注入`uigf_gacha_type`，以备双池子之需

我记得原始数据的那个已经可以区分双池了

> 3\. 我认为应该注入新的`uigf_item_type`属性而不是修改原始数据

如果SR保持现有API格式，那就没有必要，主要是GI在刚开服时 item_id 同样是包含真实值的

> 4\. 加个`game`还是很有必要的

如果单开一个新的不同名称的spec就不需要这个了，如果最后依旧是升级UIGF的话那确实需要类似的identifier

**Reply by Lightczx** [reply]
@TremblingMoeNew  UISRGF（🤣）

**Reply by TremblingMoeNew** [reply]
> > 3. 和UIGF.J v2.1一样向记录项中注入`uigf_gacha_type`，以备双池子之需
> 
> 我记得原始数据的那个已经可以区分双池了

从[Dimbreath/StarRailData](https://github.com/Dimbreath/StarRailData)看，原始数据里那个2003 2004是希儿角色卡池 希儿光锥卡池，等到了景元池子编号又变了……

可能还是得防一手301与400之分

现在不加、真加了再说也行就是了

**Reply by Lightczx** [reply]
所以卡池类型目前有对应的id么

**Reply by DancingSnow0517** [reply]
> 所以卡池类型目前有对应的id么

应该是有的，api有返回个`gacha_id`，到时候就看看景元池这个是啥

**Reply by TremblingMoeNew** [reply]
> 所以卡池类型目前有对应的id么

`gacha_type`还是，1是常驻2是新手11是角色12是光锥

我刚才说的是`gacha_id`。主要是不知道如果开双池子，那会是同样的type不同的id还是连type都和原神一样换了

**Reply by TremblingMoeNew** [reply]
> > 所以卡池类型目前有对应的id么
> 
> 应该是有的，api有返回个`gacha_id`，到时候就看看景元池这个是啥

……你用的`gacha_id`区分的？

**Reply by DancingSnow0517** [reply]
> > > 所以卡池类型目前有对应的id么
> > 
> > 
> > 应该是有的，api有返回个`gacha_id`，到时候就看看景元池这个是啥
> 
> ……你用的`gacha_id`区分的？

那肯定是`gacha_type`, `gacha_id`以后可以用去每个卡池的抽卡分析吧

**Reply by MoRanYue** [reply]
> > > > 所以卡池类型目前有对应的id么
> > > 
> > > 
> > > 应该是有的，api有返回个`gacha_id`，到时候就看看景元池这个是啥
> > 
> > 
> > ……你用的`gacha_id`区分的？
> 
> 那肯定是`gacha_type`, `gacha_id`以后可以用去每个卡池的抽卡分析吧

`gacha_id`竟然有用，我还以为它没用呢

**Reply by AuroraZiling** [reply]
有关`gacha_id`:
```
{
      "count": "1",
      "gacha_id": "2003",
      "gacha_type": "11",
      "id": "1682845800003953620",
      "item_id": "1102",
      "item_type": "角色",
      "name": "希儿",
      "rank_type": "5",
      "time": "2023-04-30 17:18:41",
      "uigf_gacha_type": "11"
},
{
      "count": "1",
      "gacha_id": "2004",
      "gacha_type": "11",
      "id": "1684318200015555520",
      "item_id": "1204",
      "item_type": "角色",
      "name": "景元",
      "rank_type": "5",
      "time": "2023-05-17 18:01:06",
      "uigf_gacha_type": "11"
}
```

**Comment by Lightczx** [comment]
新 spec 就叫 `SRGF` 吧 
谁赞成，谁反对？

**Reply by Scighost** [reply]
Star Rail Gacha Format?

**Reply by Lightczx** [reply]
> Star Rail Gacha Format?

Star Rail GachaLog Format

对标 Uniformed Interchangeable GachaLog Format

**Reply by Masterain98** [reply]
那 `UI` 的含义就没了

**Reply by Lightczx** [reply]
叫 UISRGF 也可以

**Reply by TremblingMoeNew** [reply]
我觉得SRGF可以

**Reply by ghost** [reply]
赞同 SRGF

**Reply by LaoshuBaby** [reply]
> > Star Rail Gacha Format?
> 
> Star Rail GachaLog Format
> 
> 对标 Uniformed Interchangeable GachaLog Format

~~那“UI”就会显得很名不副实，明明只是原的格式，却顶上了Uniformed的字样，在只有原的时候自然没啥问题~~

**Reply by ghost** [reply]
1. 格式缩写应该尽量简洁明了，6个字母稍微有点长
2. `UI` 字母我认为去掉也不影响他的语义和实际作用，推动让各个程序都接受的话自然就是 `UI` 了

**Reply by Masterain98** [reply]
其实具体标准的名称是随意的，可以把 `UIGF-Org` 中的 `G` 理解为 `Game` 就好了（

**Reply by zzpro** [reply]
SRGF，可行

**Comment by TremblingMoeNew** [comment]
`region`有没有保留的需要🤔

**Reply by Masterain98** [reply]
> `region`有没有保留的需要🤔

如果能通过UID来判断服务器就不需要。原神中，每个服务器有统一的时区标准；星轨情况是否一致需要调查

**Reply by TremblingMoeNew** [reply]
> > `region`有没有保留的需要🤔
> 
> 如果能通过UID来判断服务器就不需要。原神中，每个服务器有统一的时区标准；星轨情况是否一致需要调查

`region_time_zone`我觉得留着比较方便

**Reply by ghost** [reply]
@TremblingMoeNew  请问 `region_time_zone` 是具体有什么作用呢，和 `game_biz` 有映射关系吗？

如果不从抽卡链接返回的结果拿这个值，是否可以从其他方式获取或者生成呢？

**Reply by DancingSnow0517** [reply]
> @TremblingMoeNew  请问 `region_time_zone` 是具体有什么作用呢，和 `game_biz` 有映射关系吗？
> 
> 如果不从抽卡链接返回的结果拿这个值，是否可以从其他方式获取或者生成呢？
> 

api给的time是一个格式化的时间字符串，而且没有时区信息，需要region_time_zone来得到正确的时间戳，以免出现时间错位

**Reply by TremblingMoeNew** [reply]
> @TremblingMoeNew 请问 `region_time_zone` 是具体有什么作用呢，和 `game_biz` 有映射关系吗？

`region_time_zone`标识了用户所属服务器的时区。跃迁记录项里的`time`就是按服务器时间来的。在需要对跃迁记录项时间进行时区转换时就用上了~~比如我的就依赖时区数据~~。与`game_biz`没有映射关系，`hkrpg_global`又不是一个服务器，是好几个。

**Reply by DancingSnow0517** [reply]
> > @TremblingMoeNew 请问 `region_time_zone` 是具体有什么作用呢，和 `game_biz` 有映射关系吗？
> 
> `region_time_zone`标识了用户所属服务器的时区。跃迁记录项里的`time`就是按服务器时间来的。在需要对跃迁记录项时间进行时区转换时就用上了~~比如我的就依赖时区数据~~。与`game_biz`没有映射关系，`hkrpg_global`又不是一个服务器，是好几个。

我的也依赖了时区数据，不然时间的换算有点麻烦（

**Reply by ghost** [reply]
服务器间和本地可能不一致，所以需要还需要时区信息来转换为本地时间，这样理解可以吗

**Comment by Masterain98** [comment]
UIGF API 生成的星轨数据预览

[uigf_dict_sql.zip](https://github.com/UIGF-org/.github/files/11469118/uigf_dict.zip)
[uigf_dict_json.zip](https://github.com/UIGF-org/.github/files/11469121/uigf_dict_json.zip)

**Comment by LittleNyima** [comment]
插个眼，我也有个星铁抽卡导出的仓库 [LittleNyima/honkai-starrail-toolkit](https://github.com/LittleNyima/honkai-starrail-toolkit)，什么时候统一标准制定好的时候求踢我一脚，我也接入一下（）

**Comment by Lightczx** [comment]
目前我建议的格式：

```
{
  "info": {
    "uid": "xxxxxxxxx",
    "lang": "zh-cn",
    "region_time_zone": "xxxx",
    "export_timestamp": 1684124992,
    "export_app": "xxx",
    "export_app_version": "xxxxx",
    "srgf_version": "v1.0"
  },
  "list": [
    {
      "gacha_id": "2003",
      "gacha_type": "200",
      "item_id": "10000065",
      "count": "1",
      "time": "2023-05-07 10:47:00",
      "name": "xxx",
      "item_type": "xxxx",
      "rank_type": "4",
      "id": "1683425160001357994"
    }
  ]
}
```

类似 srgf_gacha_type的字段暂时不添加吧

**Reply by TremblingMoeNew** [reply]
`region_time_zone` 我认为应该加进`info`里

**Reply by Lightczx** [reply]
> `region_time_zone` 我认为应该加进`info`里

加了

**Comment by Bill-Haku** [comment]
插个眼，我们的App [星铁披萨小助手 / HSR Pizza Helper](https://github.com/pizza-studio/hsrpizzahelper)也会支持新标准

**Comment by AuroraZiling** [comment]
++插个眼，[star-rail-asta](https://github.com/AuroraZiling/star-rail-asta)项目也支持

**Comment by 5yuefengyang** [comment]
非小酋前来报道

**Comment by Lightczx** [comment]
@UIGF-org/srgf-update-audience https://github.com/UIGF-org/UIGF-org.github.io/pull/16

**Comment by Lightczx** [comment]
https://uigf.org/standards/SRGF.html
